### ISSUE TITLE
Transmission 4.0.2 freezes when quitting the app on macOS

### ISSUE THREAD

**Issue by Jerry23011** [transmission/transmission#5243]
### Steps to reproduce
1. Add a torrent through magnet
2. Choose to Remove and trash data file halfway through the download
3. Quit Transmission
4. Transmission takes a long time quitting and the cursor turns to the beach ball

### Which application of Transmission?

macOS app

### Which version of Transmission?

4.0.2 (2a57b17031)

**Comment by nevack** [comment]
> The GTK client has a little popup saying "Sending upload/download totals to tracker…"; maybe that should be present in the macOS client too.

Maybe we could add some spinner with message, blocking all other UI (and shortcust), but not blocking UI thread.
This way App will not freeze, but will be unaccessible.

Also, some time ago Transmission displayed "Quitting" badge in Dock, dunno why this was removed.

**Comment by ckerr** [comment]
Good point. I should've been clearer & said that the popup in the GTK client is modal so that the rest of the client's UI is inaccessible.

**Comment by GaryElshaw** [comment]
This isn't a new issue, it's related to the process sequencing when macOS is quit. Possibly also related #4759

**Comment by Jerry23011** [comment]
Hi, everyone, thanks for looking into this.
> This isn't a new issue, it's related to the process sequencing when macOS is quit. Possibly also related https://github.com/transmission/transmission/issues/4759

I just checked the screen recording in #4759, it looks very similar to the issue I'm encountering. I don't need to force quit, though. Transmission just freezes for a few seconds and quit itself.

**Comment by ckerr** [comment]
@GaryElshaw agreed, 4759 is another possible cause.

@Jerry23011 if macOS generates a hang report, could you attach it to this ticket?

**Comment by Jerry23011** [comment]
> @Jerry23011 if macOS generates a hang report, could you attach it to this ticket?

I'd love to, but macOS hasn't generated any reports yet.

**Comment by Coeur** [comment]
Transmission nightlies have been freezing on quit for 10 seconds ever since I started using them, more than a year ago. But that is tracked by https://github.com/transmission/transmission/issues/4759.

**Comment by frozenpandaman** [comment]
Happening to me too – usually just takes ~10 seconds to close.

Sometimes, however, it completely hangs, requires a Force Quit, and after performing this, it disappears from Activity Monitor but still appears in the Force Quit menu… which necessitates a full computer restart (!)

**Comment by Coeur** [comment]
> Also, some time ago Transmission displayed "Quitting" badge in Dock, dunno why this was removed.

@nevack see #2495

**Comment by nevack** [comment]
> > Also, some time ago Transmission displayed "Quitting" badge in Dock, dunno why this was removed.
> 
> @nevack see #2495

Thanks, checked the issue #1690 for motivation.
From UX point of view app is clearly hanging on quit.
I think we would need to implement another way of indication.

**Comment by G7TOK** [comment]
Hi,
I sometimes have to wait 30 seconds for the app to quit, I often force quit.
This is top part of the Sequoia problem report.

**Heaviest stack for the main thread of the target process:
  17  start + 2840 (dyld + 25204) [0x19007c274]
  17  ??? (Transmission + 9656) [0x1025d25b8]
  17  NSApplicationMain + 888 (AppKit + 18516) [0x194016854]
  17  -[NSApplication run] + 480 (AppKit + 188512) [0x194040060]
  17  -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688 (AppKit + 10089172) [0x1949b12d4]
  17  _DPSNextEvent + 660 (AppKit + 241716) [0x19404d034]
  17  _BlockUntilNextEventMatchingListInModeWithFilter + 76 (HIToolbox + 1119496) [0x19ba40508]
  17  ReceiveNextEventCommon + 216 (HIToolbox + 1118588) [0x19ba4017c]
  17  RunCurrentEventLoopInMode + 292 (HIToolbox + 1094960) [0x19ba3a530]
  17  CFRunLoopRunSpecific + 588 (CoreFoundation + 505636) [0x1904e2724]
  17  __CFRunLoopRun + 824 (CoreFoundation + 508184) [0x1904e3118]
  17  __CFRunLoopDoBlocks + 356 (CoreFoundation + 512816) [0x1904e4330]
  17  __CFRUNLOOP_IS_CALLING_OUT_TO_A_BLOCK__ + 28 (CoreFoundation + 513056) [0x1904e4420]
  17  ___NSMainRunLoopPerformBlockInModes_block_invoke + 44 (AppKit + 6010232) [0x1945cd578]
  17  __52-[NSApplication(NSAppleEventHandling) _handleAEQuit]_block_invoke + 52 (AppKit + 3572900) [0x19437a4a4]
  17  -[NSApplication _terminateFromSender:askIfShouldTerminate:saveWindows:] + 124 (AppKit + 3573176) [0x19437a5b8]
  17  -[NSApplication terminate:] + 1660 (AppKit + 2371160) [0x194254e58]
  17  -[NSNotificationCenter postNotificationName:object:userInfo:] + 88 (Foundation + 40916) [0x191661fd4]
  17  _CFXNotificationPost + 732 (CoreFoundation + 266140) [0x1904a7f9c]
  17  _CFXRegistrationPost + 436 (CoreFoundation + 1061204) [0x19056a154]
  17  ___CFXRegistrationPost_block_invoke + 88 (CoreFoundation + 1061388) [0x19056a20c]
  17  __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148 (CoreFoundation + 467808) [0x1904d9360]
  17  ??? (Transmission + 30272) [0x1025d7640]
  17  ??? (Transmission + 685488) [0x1026775b0]
  17  std::__1::__assoc_sub_state::wait() + 56 (libc++.1.dylib + 138648) [0x190332d98]
  17  std::__1::__assoc_sub_state::__sub_wait(std::__1::unique_lock<std::__1::mutex>&) + 56 (libc++.1.dylib + 138484) [0x190332cf4]
  17  std::__1::condition_variable::wait(std::__1::unique_lock<std::__1::mutex>&) + 28 (libc++.1.dylib + 136568) [0x190332578]
  17  __psynch_cvwait + 8 (libsystem_kernel.dylib + 18156) [0x1903be6ec]
 *17  psynch_cvcontinue + 0 (com.apple.kec.pthread + 17956) [0xfffffe000bb4f414]**

I can send the whole thing if needed.

Gary

**Comment by Coeur** [comment]
>I can post the whole thing if needed.

No need. Your stack trace isn't symbolised, it's all with "???":
> 17 ??? (Transmission + 9656) [0x1025d25b8]

And it's 100% reproducible by all devs already: devs with time on their hands can already look into it.

**Comment by qacizar** [comment]
What's the solution?

**Comment by seakrebel** [comment]
Any solution yet?

**Comment by Coeur** [comment]
As a reminder, the initial scope of this issue is a UI issue described by ckerr and consists of adding:
> a little popup saying "Sending upload/download totals to tracker…"

If you want the app to quit faster, it might be tracked by a different issue:
> that is tracked by [#4759](https://github.com/transmission/transmission/issues/4759) at the moment (previously was tracked by [#2705](https://github.com/transmission/transmission/issues/2705)).
